File: GitToolBox/src/main/java/zielu/gittoolbox/cache/LazyRepoChangeAware.java
package zielu.gittoolbox.cache;

import git4idea.repo.GitRepository;
import java.util.function.Supplier;
import org.jetbrains.annotations.NotNull;
import zielu.gittoolbox.util.MemoizeSupplier;

class LazyRepoChangeAware<T extends RepoChangeAware> implements RepoChangeAware {
  private final Supplier<T> supplier;

  LazyRepoChangeAware(Supplier<T> delegateSupplier) {
    supplier = new MemoizeSupplier<>(delegateSupplier);
  }

  @Override
  public void repoChanged(@NotNull GitRepository repository) {
    supplier.get().repoChanged(repository);
  }
}

File: src/main/kotlin/zielu/gittoolbox/cache/LazyRepoChangeAware.kt
package zielu.gittoolbox.cache

import git4idea.repo.GitRepository

internal class LazyRepoChangeAware<out T : RepoChangeAware> (
  supplier: () -> T
) : RepoChangeAware {
  private val delegate: T by lazy {
    supplier.invoke()
  }

  override fun repoChanged(repository: GitRepository) {
    delegate.repoChanged(repository)
  }
}
